### PR TITLE
vc: Detach device when unable to store sandbox device

### LINKED ISSUE #2301: Detach device when unable to store sandbox device
# Description of problem

In Container#mountSharedDirMounts, if sandbox.storeSandboxDevices() returns error, we should detach the device.

# Expected result

This addresses existing comment:
```
                               if err := c.sandbox.storeSandboxDevices(); err != nil {
                                       //TODO: roll back?
```
# Actual result

### PR DESCRIPTION
**Pull request by tedyu**
In Container#mountSharedDirMounts, if sandbox.storeSandboxDevices() returns error, we should detach the device.

Fixes #2301

Signed-off-by: Ted Yu yuzhihong@gmail.com

### CODE REVIEW

File: virtcontainers/container.go [hunk 1]
```diff
@@ -574,7 +574,8 @@ func (c *Container) mountSharedDirMounts(hostSharedDir, guestSharedDir string) (
```
Comment (bergwolf) [inline virtcontainers/container.go]: Better adding a defer rollback function here for each devices attached.

Comment (tedyu) [reply virtcontainers/container.go]: The detach is done on line 577.

**Comment by tedyu** [comment]
@bergwolf 
Please take another look.

File: virtcontainers/container.go [hunk 2]
```diff
@@ -571,10 +580,12 @@ func (c *Container) mountSharedDirMounts(hostSharedDir, guestSharedDir string) (
 			if err := c.sandbox.devManager.AttachDevice(m.BlockDeviceID, c.sandbox); err != nil {
 				return nil, nil, err
 			}
+			devicesToDetach = append(devicesToDetach, m.BlockDeviceID)
 
 			if !c.sandbox.supportNewStore() {
-				if err := c.sandbox.storeSandboxDevices(); err != nil {
-					//TODO: roll back?
+				if err = c.sandbox.storeSandboxDevices(); err != nil {
+					c.sandbox.devManager.DetachDevice(m.BlockDeviceID, c.sandbox)
```
Comment (bergwolf) [inline virtcontainers/container.go]: not needed any more? It's included in devicesToDetach IIUC.

**Comment by tedyu** [comment]
@bergwolf 
Please take another look.

**Comment by WeiZhang555** [comment]
`storeSandboxDevices` function call is removed in https://github.com/kata-containers/runtime/pull/2239 .
If so, do we still need this PR?

**Comment by tedyu** [comment]
There is other return with err:
```
			return nil, nil, err
```
So we still need this PR.

**Comment by tedyu** [comment]
@jodh-intel @grahamwhaley
Mind triggering ubuntu test run ?

thanks

**Comment by tedyu** [comment]
In favor of #2308

**Pull request closed without merging**